### PR TITLE
WIP: Quick experiment for exposing DI to plugins

### PR DESCRIPTION
**Pull request by juho-jaakkola**
I made a quick test of minimal changes needed to expose DI API to plugins.

Any opinions so far? What do we have to take into consideration when exposing the API?

There was earlier at least some concern about exposing core services to plugins because this complicates refactoring. This is why we have the `\Elgg\Application::public_services` property, that controls which of the core services can be accessed outside core. Then again I would really like to expose as many of the core services as possible.

Some related issues:
 - #5930 - Expose DI system to plugins
 - #7461 - OO Event + plugin hook handlers
 - #6387 - Use PHP-DI for our dependency injector

### CODE REVIEW

File: engine/classes/Elgg/Application.php [hunk 1]
```diff
@@ -320,6 +320,13 @@ public function __get($name) {
 		trigger_error("Undefined property: " . __CLASS__ . ":\${$name}");
 	}
 
+	/**
+	 *
+	 */
+	public function service($name) {
+		return $this->_services->{$name};
+	}
```
Comment (juho-jaakkola) [inline engine/classes/Elgg/Application.php]: I added this just to temporarily bypass the `public_services` property and getting access to all services.

File: mod/blog/classes/Elgg/Blog/Bar.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+	private $foo;
+
+	public function __construct(Foo $foo) {
+		$this->foo = $foo;
+	}
+
+	public function test ($test) {
+		$this->foo->test("Calling Bar::test() with value: $test");
+	}
```
Comment (juho-jaakkola) [inline mod/blog/classes/Elgg/Blog/Bar.php]: Example of a plugin injecting its own service to another.

File: mod/blog/classes/Elgg/Blog/Foo.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+	private $system_messages;
+
+	public function __construct(SystemMessagesService $system_messages) {
+		$this->system_messages = $system_messages;
+	}
+
+	public function test($param) {
+		$this->system_messages->addSuccessMessage($param);
+	}
```
Comment (juho-jaakkola) [inline mod/blog/classes/Elgg/Blog/Foo.php]: Example of a plugin injecting a core service.

File: engine/lib/constants.php [hunk 4]
```diff
@@ -155,6 +155,12 @@
  */
 define('ELGG_PLUGIN_IGNORE_MANIFEST', 64);
 
+
+/**
+ * Tells \ElggPlugin::start() to automatically register the plugin's services.
+ */
+define('ELGG_PLUGIN_REGISTER_SERVICES', 123);
```
Comment (jeabakker) [inline engine/lib/constants.php]: should this be `128`

**Comment by jeabakker** [comment]
What is this trying to add? If you wish for more of your plugin to be testable a lot of things can be made into integration tests which has complete core available

File: engine/lib/constants.php (hunk 4 above)
Comment (juho-jaakkola) [reply engine/lib/constants.php]: I just entered something randomly.

**Comment by juho-jaakkola** [comment]
The point is to allow making PHPUnit tests for the plugins. Injecting all dependencies through the constructor allows replacing them with [test doubles](https://phpunit.de/manual/current/en/test-doubles.html) in the PHPUnit tests.

**Comment by juho-jaakkola** [comment]
There's more info in the Slack channel.

**Comment by hypeJunction** [comment]
I was just thinking about it today. I think it would be a great addition. Plugins like system log, could build their services that would run from singletons instead of rebuilding them and using stupid function with unreusable spaghetti code.

**Comment by hypeJunction** [comment]
I don't think we should expose non-public services. The core is fully mocked for testing, so there is no reason for anyone needing to access `_elgg_services()` from within their services - they should be using public APIs.
I don't know why we are so stubborn about having only one service provider - it would make a lot more sense to break it down into parts, it has become a mamoth that is so couple together, that it's a maze, with almost circular dependencies.
Public Services need to be moved out `Application` into their own container, and plugins can then just extend that public container, no need to access internal services. `Application` should only be responsibe for booting the application.

**Comment by juho-jaakkola** [comment]
@hypeJunction What do you mean by using public API instead of using core services?

If my plugin has a class that needs for example:
 - Access to database
 - Possibility to add system messages
 - Possibility to translate stuff

Then I need to pass those three as services into the class through the constructor in order to make PHPUnit tests where I can replace them with test doubles. Using global functions `elgg_get_entities()`, `system_message()` and `elgg_echo()` within the class instead would prevent me from doing that.

**Comment by mrclay** [comment]
The big question is whether to provide 1 container with everyone's dependencies, or single per-plugin containers. I think I'm now for the latter.

I think the static analysis provided by documented return types is important enough to ask plugins  to provide their own container class (with typed properties) and a function to retrieve it:

```php
function get_hello_container(): Hello\Container {
    // instantiates and caches container
    return elgg_get_container(Hello\Container::class);
}

get_hello_services()->world; // Hello\World component
```

I'm not in favor of building our own factory scheme into elgg-plugin. I think in practice you almost always need code, and if we want a static configuration I'd rather use something established and 3rd party.

**Comment by mrclay** [comment]
@juho-jaakkola Right, we're banging our head against the fact that our API is completely global, which just infects everything it touches. But I kinda cringe thinking of making 2 ways to do nearly everything in the system. But that may be where we are. Or indeed killing dead every single existing plugin to force everyone into the OO world.

**Comment by mrclay** [comment]
Well, we *can* look closer at what Laravel does, as it apparently makes testing easy even with static methods.

**Comment by juho-jaakkola** [comment]
I'm not sure if I understand what you mean by per-plugin containers. Would this mean that plugin A could not inject services provided by plugin B at all?

**Comment by mrclay** [comment]
> I'm not sure if I understand what you mean by per-plugin containers. Would this mean that plugin A could not inject services provided by plugin B at all?

Well, plugin B could trigger a hook in its container constructor/factory.

I guess the greater industry seems OK with using one shared container and trusting each plugin to behave well, but that forces every developer to add `/* @var Foo $foo */` after every dependency fetch for proper static analysis. I guess as long as plugins are free to use their own typed containers if they wish, I'm OK with whatever we do.

Strawman: 
- Shared [PHP-DI](http://php-di.org/) container for plugins and *select* core services.
- Access to the container instance via `elgg_get_container()`. This allows full access to [PHP-DI's features](http://php-di.org/doc/container.html).
- Plugin file `elgg-di.php` returns a PHP-DI definition array, they all get merged in plugin order and cached with PHP-DI's built-in caching. That means no use of `new` and similar code in that file. If you need code, you can access the container directly in `start.php`.
- We encourage using interface classnames as keys for all objects, and the use of `$container->call(function (FooInterface $foo) { ... })` which bypasses the whole concern with `get()` lacking return types.

**Comment by mrclay** [comment]
Take a look at #11546

**Comment by jeabakker** [comment]
replaced by #11674

**Pull request closed without merging**